File: Binary-Search/nth-root-of-m.py
# Time  : O(logn)
# Space : O(1)

def NthRoot(n: int, m: int) -> int:
  lo, hi = 1, m

  while lo <= hi:
    mid = lo + (hi - lo) // 2
    if mid ** n < m:
      lo = mid + 1
    elif mid ** n > m:
      hi = mid - 1
    else:
      return mid
      
  return -1

File: Binary-Search/nth-root-of-m.cpp
// Time  : O(logm)
// Space : O(1)
class Solution{
public:
	int NthRoot(int n, int m) {
    int lo = 1;
    int hi = m;
    while (lo <= hi) {
      int mid = (lo + hi)>>1;
      if (pow(mid, n) == m) return mid;
      else if (pow(mid, n) < m) lo = mid+1;
      else hi = mid-1;
    }
    return -1;
	}
};